File: Math/13. Roman to Integer_E.c

int LtoInt(char c){
    
    switch (c){
        case 'I' : return 1;
        case 'V' : return 5;
        case 'X' : return 10;
        case 'L' : return 50;
        case 'C' : return 100;
        case 'D' : return 500;
        case 'M' : return 1000;
        default: return 0;
    }
    
    
}

/*
[approach] we iterate the string char one by one and add the value of each char to the sum.
If we see a char whose value is smaller than the last char's value, we substact the sum by last value times 2.
That is how we make it right.
time_O(N) we iterate the string once.
spact_O(1) we only need to store last char's value.

*/
int romanToInt(char * s){
    char * p = NULL;
    int curInt = 0;
    int lastInt = 0;
        
    int output = 0;
    if (s == NULL){
        return 0;
    }
    
    p = s;
    for ( p = s; *p!= '\0'; p++){
        curInt = LtoInt(*p);
        output = output + curInt;
        if (lastInt != 0 && lastInt < curInt){
            output = output - 2* lastInt;
        }         
        lastInt = curInt;  
    }
    return output;
}

File: Math/13. Roman to Integer_E.py
class Solution:
    def romanToInt(self, s: str) -> int:
        """
        time: 08152022  8:30---
        [Approach]
        It is a string , we can iterate it, see how many each letter stands for.
        We also need to keep a order, when smaller letter appears before bigger letter, we treats it differently.
        time_O(N) go through the string once.
        spaceO(1) no extra space needed.
        """
        
        LtoInt = {
            "I" : 1,
            "V" : 5,
            "X" :10,
            "L"  : 50,
            "C"  : 100,
            "D"  : 500,
            "M"  : 1000,
            "IV" : 4,
            "IX"  :9,
            "XL"  : 40,
            "XC"  :90,
            "CD"  : 400,
            "CM"   :900
         }
        
        sSet = {"I", "X", "C"}
        
        output = 0
        
        l = len(s)
        i = 0
        while i < l:
            if s[i] in sSet and i < l-1 and s[i:i+2] in LtoInt:
                output += LtoInt[ s[i:i+2] ]
                i += 2
            else:
                output += LtoInt[ s[i] ]
                i += 1
        return output
                    